Trying to address an incremental compilation issues

pacak opened the pull request:
This pull request contains two independent changes, one makes it so when `try_force_from_dep_node` fails to recover a query - it marks the node as "red" instead of "green" and the second one makes Debug impl for `DepNode` less panicky if it encounters something from the previous compilation that doesn't map to anything in the current one.

I'm not 100% confident that this is the correct approach, but so far I managed to find a bunch of comments suggesting that some things are allowed to fail in a certain way and changes I made are allowing for those things to fail this way and it fixes all the small reproducers I managed to find.

Compilation panic this pull request avoids is caused by an automatically generated code on an associated type and it is not happening if something else marks it as outdated first (or close like that, but scenario is quite obscure).

Fixes https://github.com/rust-lang/rust/issues/107226
Fixes https://github.com/rust-lang/rust/issues/125367

pacak committed:
Allow for try_force_from_dep_node to fail

The way it is implemented currently try_force_from_dep_node returns true
as long as there's a function to force the query. It wasn't this way
from the beginning, earlier version was producing forcing result and it
was changed in https://github.com/rust-lang/rust/pull/89978, I couldn't
find any comments addressing this change.

One way it can fail is by failing to recover the query in
DepNodeParams::recover - when we are trying to query something that no
longer exists in the current environment

pacak committed:
local_def_path_hash_to_def_id can fail

local_def_path_hash_to_def_id is used by Debug impl for DepNode and it
looks for DefPathHash inside the current compilation. During incremental
compilation we are going through nodes that belong to a previous
compilation and might not be present and a simple attempt to print such
node with tracing::debug (try_mark_parent_green does it for example)
results in a otherwise avoidable panic

Panic was added in https://github.com/rust-lang/rust/pull/82183,
specifically in 2b60338ee9, with a comment "We only use this mapping for
cases where we know that it must succeed.", but I'm not sure if this
property holds when we traverse nodes from the old compilation in order
to figure out if they are valid or not

michaelwoerister commented:
Thanks for the PR, @pacak! I'll take a look asap.

pacak commented:
Looking at https://github.com/rust-lang/rust/issues?q=is%3Aissue+is%3Aopen+label%3AA-incr-comp - I think this pull request should help to close quite a few of open incremental compilation issues. I'll go though the list later.

michaelwoerister commented:
The `try_force_from_dep_node` issue definitely looks like a bug that has been introduced in the commit that you point to. That's an awesome find, @pacak! Thanks for investigating this.

I'm wondering if there is an obvious way to write a test for this.

michaelwoerister commented on compiler/rustc_query_system/src/dep_graph/mod.rs:
```diff
@@ -56,15 +56,15 @@ pub trait DepContext: Copy {
+                    panic::resume_unwind(value)
                 }
-                panic::resume_unwind(value)
+                Ok(val) => val,
```
```suggestion
                Ok(query_has_been_forced) => query_has_been_forced,
```

michaelwoerister commented on compiler/rustc_query_system/src/dep_graph/mod.rs:
```diff
@@ -56,15 +56,15 @@ pub trait DepContext: Copy {
     fn try_force_from_dep_node(self, dep_node: DepNode, frame: Option<&MarkFrame<'_>>) -> bool {
```
Can you add something like this to the doc comment:

> Returns `true` if the query has actually been forced. It is valid that a query fails to be forced, e.g. when the query key cannot be reconstructed from the dep-node or when the query kind outright does not support it.

michaelwoerister commented:
Your test case from https://github.com/rust-lang/rust/issues/107226#issuecomment-2159004254 seems good to add.

Making the `DefPathHash -> (Local)DefId` mappings return an `Option` seems good, especially with respect to not crashing the `DepNode` debug impl.

I'm wondering about some of the error stack traces though, e.g. [here](https://github.com/rust-lang/rust/issues/125945#issue-2332227862). Normally, we should not encounter a removed `DefPathHash` during red-green-marking because at that point we should have come upon a red node signifying the removal and thus stopped going any further. But in these backtraces we do encounter these nodes and  the query traces all have something with trait-solving in them. Trait-solving is the only area where we do manual dependency tracking and I'm wondering if this points to a bug in that manual tracking or if gracefully failing to reconstruct the dep-node is the valid way to handle this.

pacak commented:
There's probably a bug sitting somewhere, but not having random panic when
you try to print stuff makes sense. If anything - we can look into adding
an assertion that traversed nodes are safe with some better error if they
are not.

On Mon, Jun 17, 2024, 10:38 Michael Woerister ***@***.***>
wrote:

> Your test case from #107226 (comment)
> <https://github.com/rust-lang/rust/issues/107226#issuecomment-2159004254>
> seems good to add.
>
> Making the DefPathHash -> (Local)DefId mappings return an Option seems
> good, especially with respect to not crashing the DepNode debug impl.
>
> I'm wondering about some of the error stack traces though, e.g. here
> <https://github.com/rust-lang/rust/issues/125945#issue-2332227862>.
> Normally, we should not encounter a removed DefPathHash during
> red-green-marking because at that point we should have come upon a red node
> signifying the removal and thus stopped going any further. But in these
> backtraces we do encounter these nodes and the query traces all have
> something with trait-solving in them. Trait-solving is the only area where
> we do manual dependency tracking and I'm wondering if this points to a bug
> in that manual tracking or if gracefully failing to reconstruct the
> dep-node is the valid way to handle this.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/126409#issuecomment-2173598217>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAQFI5FICFR2RM4MPGM62DZH3YHBAVCNFSM6AAAAABJINBXHSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNZTGU4TQMRRG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

michaelwoerister commented:
> but not having random panic when you try to print stuff makes sense.

Agreed 🙂

pacak commented:
I rebased on top of master, addressed your suggestions to naming/comments, added a test and confirmed that the test fails before changes to `try_force_from_dep_node` and passes after.

At some point I'll look into requests in trait solving code, but it might take time.

michaelwoerister commented on compiler/rustc_hir/src/definitions.rs:
```diff
@@ -377,22 +377,12 @@ impl Definitions {
-        hash: DefPathHash,
-        err_msg: &dyn std::fmt::Debug,
-    ) -> LocalDefId {
+    pub fn local_def_path_hash_to_def_id(&self, hash: DefPathHash) -> Option<LocalDefId> {
```
Please add a doc string:
```rust
/// Returns `None` if the `DefPathHash` does not correspond to a `LocalDefId` 
/// in the current compilation session. This can legitimately happen if the 
/// `DefPathHash` is from a `DefId` in an upstream crate or, during incr. comp.,
/// if the `DefPathHash` is from a previous compilation session and
/// the def-path does not exist anymore.
```

pacak replied:
Done.

michaelwoerister reviewed (commented):
Thank you, @pacak, this is a great find! Would you mind adding the doc comment suggested below? Then we can merge this.

michaelwoerister commented:
> At some point I'll look into requests in trait solving code, but it might take time.

Thank you! That would be great. Please ping me when you do so.

pacak commented:
> Would you mind adding the doc comment suggested below? Then we can merge this.

Done. I believe this pull request should close a lot of existing incremental compilation tickets, I listed two of them but there's more and they don't have reproducers so I wasn't able to test them. Do you think they can be closed as well? I don't have a concrete list, but going though tickets labeled with incremental compilation ICEs and looking at stack traces shows them. 

> Thank you! That would be great. Please ping me when you do so.

Sure.

michaelwoerister commented:
> Do you think they can be closed as well?

Let's merge the PR and once it is in nightly, ask for confirmation on the respective issues.

michaelwoerister commented:
Let's not roll this up, for better bisectability.
@bors r+ rollup=never

bors commented:
:pushpin: Commit 12f8d12b418c9bbe4d2142f2c8e00a28a5b6dfe5 has been approved by `michaelwoerister`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=michaelwoerister 12f8d12b418c9bbe4d2142f2c8e00a28a5b6dfe5 -->
<!-- homu: {"type":"Approved","sha":"12f8d12b418c9bbe4d2142f2c8e00a28a5b6dfe5","approver":"michaelwoerister","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 12f8d12b418c9bbe4d2142f2c8e00a28a5b6dfe5 with merge 1d96de2a20e963abb8923dfa3c6175517dfb9d2c...
<!-- homu: {"type":"BuildStarted","head_sha":"12f8d12b418c9bbe4d2142f2c8e00a28a5b6dfe5","merge_sha":"1d96de2a20e963abb8923dfa3c6175517dfb9d2c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9594701821/job/26463581905)
Approved by: michaelwoerister
Pushing 1d96de2a20e963abb8923dfa3c6175517dfb9d2c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"michaelwoerister","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9594701821/job/26463581905"},"merge_sha":"1d96de2a20e963abb8923dfa3c6175517dfb9d2c"} -->

rust-timer commented:
Finished benchmarking commit (1d96de2a20e963abb8923dfa3c6175517dfb9d2c): [comparison URL](https://perf.rust-lang.org/compare.html?start=1208eddaff59c98e37574c860de6f68aa404958f&end=1d96de2a20e963abb8923dfa3c6175517dfb9d2c&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=1208eddaff59c98e37574c860de6f68aa404958f&end=1d96de2a20e963abb8923dfa3c6175517dfb9d2c&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.3% | [-0.4%, -0.2%] | 21    |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.3%, -0.2%] | 15    |
| All ❌✅ (primary)                 | -0.3% | [-0.4%, -0.2%] | 21    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1208eddaff59c98e37574c860de6f68aa404958f&end=1d96de2a20e963abb8923dfa3c6175517dfb9d2c&stat=max-rss)
<details>
<summary>Results (primary 3.0%, secondary 3.5%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.0% | [2.2%, 4.1%] | 3     |
| Regressions ❌ <br /> (secondary)  | 3.5% | [3.5%, 3.5%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 3.0% | [2.2%, 4.1%] | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1208eddaff59c98e37574c860de6f68aa404958f&end=1d96de2a20e963abb8923dfa3c6175517dfb9d2c&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=1208eddaff59c98e37574c860de6f68aa404958f&end=1d96de2a20e963abb8923dfa3c6175517dfb9d2c&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 690.663s -> 691.418s (0.11%)
**Artifact size**: 323.78 MiB -> 323.81 MiB (0.01%)
